AK/String.cpp

```diff
@@ -106,6 +106,18 @@ String String::empty()
     return StringImpl::the_empty_stringimpl();
 }
 
+bool String::copy_characters_to_buffer(char* buffer, size_t buffer_size) const
+{
+    // We must fit at least the NUL-terminator.
+    ASSERT(buffer_size > 0);
```

Comment:
Because I wanted to guarantee that even when it `return`s `false`, the buffer is left in a consistent state — namely, it contains the truncated but properly null-terminated string. So you can still pass it to somewhere where a C string is expected and have it "work" (as much as it can given the string got truncated). In this light, the "yet another case that one has to consider" is "what if the string ended up not being null-terminated because the buffer was empty in the first place", and I'd rather we catch that straight away.

That being said, your logic makes sense too. If you care enough, we can switch it around.

> Cases where this might remotely make sense: We're assembling a larger buffer out of "stuff", and somehow have exactly 0 bytes left, which so far is not yet a problem.

No, if you're assembling a buffer (so basically, catting together some strings), you're going to want to overwrite last part's 0-terminator each time, so that the strings are directly adjacent. So you always have at least one byte of space.